### PR TITLE
[SPARK-42975][SQL] Cast result type to timestamp type for string +/- interval

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR makes it cast the result type of string +/- interval to timestamp type instead of string type.

### Why are the changes needed?

Both of them should evaluate to `true`. Otherwise, users will be confused:
```sql
select '2023-01-01' >= '2023-01-08' - interval '7' day;
select '2023-01-01' >= date_sub('2023-01-08', 7);
```

### Does this PR introduce _any_ user-facing change?

This PR changed the result type.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -424,6 +428,8 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
           case (DateType, CalendarIntervalType) =>
             DatetimeSub(l, r, DateAddInterval(l,
               UnaryMinus(r, mode == EvalMode.ANSI), ansiEnabled = mode == EvalMode.ANSI))
+          case (StringType, CalendarIntervalType | _: DayTimeIntervalType) =>
+            Cast(DatetimeSub(l, r, TimeAdd(l, UnaryMinus(r, mode == EvalMode.ANSI))), TimestampType)
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:432]: cc @cloud-fan @yaooqinn

**Comment by cloud-fan** [comment]
The change makes sense, but I'd say this is a legacy feature and the existing behavior doesn't make sense at all. For string +/- internal, the string can be timestamp, timestamp_ntz and interval, there is no correct behavior here.

My suggestion is don't touch it to keep legacy workloads running. We should update the SQL queries to not use String so extensively.

**Comment by cloud-fan** [comment]
Or we should probably fail it in ANSI mode, cc @gengliangwang

**Comment by wangyum** [comment]
+1 for fail it in ANSI mode.

**Comment by gengliangwang** [comment]
> My suggestion is don't touch it to keep legacy workloads running. We should update the SQL queries to not use String so extensively.

+1, totally agree!

**Pull request closed without merging**